src/SendGrid/Permissions/ISendGridPermissionScope.cs

```diff
@@ -0,0 +1,50 @@
+﻿namespace SendGrid.Permissions
+{
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// A contract for a API Key permission scope
+    /// </summary>
+    public interface ISendGridPermissionScope
+    {
+        /// <summary>
+        /// Gets the name.
+        /// </summary>
+        /// <value>
+        /// The name.
+        /// </value>
+        string Name { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether this instance is mutually exclusive.
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if this instance is mutually exclusive; otherwise, <c>false</c>.
+        /// </value>
+        bool IsMutuallyExclusive { get; }
+
+        /// <summary>
+        /// Gets the sub scopes.
+        /// </summary>
+        /// <value>
+        /// The sub scopes.
+        /// </value>
+        IEnumerable<ISendGridPermissionScope> SubScopes { get; }
+
+        /// <summary>
+        /// Gets the allowed options.
+        /// </summary>
+        /// <value>
+        /// The allowed options.
+        /// </value>
+        ScopeOptions AllowedOptions { get; }
+
+        /// <summary>
+        /// Builds the specified options.
+        /// </summary>
+        /// <param name="requestedOptions">The options.</param>
+        /// <param name="prefix">The prefix.</param>
```

Comment:
I think we might want to be a bit more descriptive here. What does a valid prefix look like?

Reply:
I'm going to refactor this a bit and just hardcode and enumerate all the scope strings per permission. This code is a few years old now and I've just re-checked and some of them have changed it looks like. Also some permissions (namely [Stats](https://sendgrid.api-docs.io/v3.0/api-key-permissions/api-key-permissions#Stats) is non-conforming to the naming of `permission.operation` and won't fit in this model. With this change this prefix will be removed.